src/middlewared/middlewared/plugins/directoryservices_/ipa_join_mixin.py

```diff
@@ -0,0 +1,419 @@
+
+        if not self.middleware.call_sync('ldap.has_ipa_host_keytab'):
+            return False
+
+        ipa_config = self.middleware.call_sync('ldap.ipa_config', ldap_conf)
+        return ipa_config['domain'].casefold() == domain.casefold()
+
+    @kerberos_ticket
+    def _ipa_set_spn(self):
+        """ internal method to create service entries on remote IPA server """
+        output = []
+        for op, spn_type in (
+            (IpaOperation.SET_SMB_PRINCIPAL, ipa_constants.IpaConfigName.IPA_SMB_KEYTAB),
+            (IpaOperation.SET_NFS_PRINCIPAL, ipa_constants.IpaConfigName.IPA_NFS_KEYTAB)
+        ):
+            try:
+                setspn = subprocess.run([
+                    IPACTL,
+                    '-a', IpaOperation.SET_SMB_PRINCIPAL.name
+                ], check=False, capture_output=True)
+            except FileNotFoundError:
+                continue
+
+            resp = _parse_ipa_response(setspn)
+            output.append(resp | {'keytab_type': spn_type})
+
+        return output
+
+    @kerberos_ticket
+    def _ipa_setup_services(self, job: Job):
+        job.set_progress(60, 'Configuring kerberos principals')
+        resp = self._ipa_set_spn()
+        domain_info = None
+
+        for entry in resp:
+            self._ipa_insert_keytab(entry['keytab_type'], entry['keytab'])
+            if entry['keytab_type'] is ipa_constants.IpaConfigName.IPA_SMB_KEYTAB:
+                domain_info = entry['domain_info'][0]
+                password = entry['password']
+
+        if domain_info:
+            job.set_progress(70, 'Configuring SMB server for IPA')
+            self.middleware.call_sync('datastore.update', 'services.cifs', 1, {
+                'cifs_srv_workgroup': domain_info['netbios_name']
+            })
+
+            # We must write the password encoded in the SMB keytab
+            # to secrets.tdb at this point.
+            self.middleware.call_sync(
+                'directoryservices.secrets.set_ipa_secret',
+                domain_info['netbios_name'],
+                base64.b64encode(password.encode())
+            )
+
+            # regenerate our SMB config to apply our new domain
+            self.middleware.call_sync('etc.generate', 'smb')
+
+            # write our domain sid to the secrets.tdb
+            setsid = subprocess.run([
+                'net', 'setdomainsid', domain_info['domain_sid']
+            ], capture_output=True, check=False)
+
+            if setsid.returncode:
+                raise CallError(f'Failed to set domain SID: {setsid.stderr.decode()}')
+
+            self.middleware.call_sync('directoryservices.secrets.backup')
+
+    @kerberos_ticket
+    def ipa_get_smb_domain_info(self) -> dict | None:
+        """
+        This information shouldn't change during normal course of
+        operations in a FreeIPA domain. Cache a copy of it for future
+        reference.
+        """
+        if self.__ipa_smb_domain is None:
+            return None
```

Comment:
Hmm.  This _might_ be fine, but just wondering whether we want `None` to be cached.  (Maybe we do.)  A comment might be useful.

Reply:
Added comment